### PR TITLE
add support for getDerivedStateFromProps

### PR DESCRIPTION
**Pull request by belaczek**

### CODE REVIEW

File: src/bindings/recompose.ts [hunk 1]
```diff
@@ -218,3 +232,90 @@ function withPropFromContext(
     }
   }
 }
+
```
Comment (belaczek) [inline src/bindings/recompose.ts:235]: I'm not sure if the whole `withState` implementations should stay here or if I should rather move it to separate file. @adam-stanek?

File: test/bindings/withState.test.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    assert.strictEqual(container.innerHTML, '<div>Hello Bob!</div>')
+  })
+
+  it('it appends state setter', async () => {
+    const decorate = createComposer()
+      .withState('salutation', 'setSalute', '')
+      .build()
+
+    const Salutation = decorate(({ salutation, setSalute }) => {
+      if (salutation === '') setSalute('Ahoj Bobe!')
```
Comment (adam-stanek) [inline test/bindings/withState.test.tsx:25]: I am not sure if this is ok. I believe that the state should not be changed inside of render method. But I have no idea how to do it better with current test solution.

File: src/bindings/recompose.ts [hunk 3]
```diff
@@ -218,3 +232,90 @@ function withPropFromContext(
...
+// type of the initialState mapper function in withState decorator
+type stateMapper<TInner, TOutter> = (input: TInner) => TOutter
```
Comment (adam-stanek) [inline src/bindings/recompose.ts]: Please name types / interfaces in PascalCase.

File: test/bindings/withState.test.tsx (hunk 2 above)
Comment (belaczek) [reply test/bindings/withState.test.tsx:25]: You are right, but I could not come up with another way how to test the derived state.

File: src/abstract.ts [hunk 4]
```diff
@@ -234,13 +234,18 @@ export interface ComponentDecoratorBuilder<
...
    * Further info: https://neoziro.github.io/recompact/#withstatestatename-stateupdatername-initialstate
    */
   withState<T3 extends string, T4 extends string, T5>(
     propName: T3,
     setterPropName: T4,
     initiaValueFromProps: (props: TResultingProps) => T5,
+    derivedStateFromProps?: (
+      props: TResultingProps,
+      state: T5,
+    ) => Partial<T5> | null,
```
Comment (adam-stanek) [inline src/abstract.ts:248]: It should return `T5 | null`

Comment (belaczek) [reply src/abstract.ts:248]: fixed

File: src/bindings/recompose.ts (hunk 1 above)
Comment (adam-stanek) [reply src/bindings/recompose.ts:235]: You can leave it as it is for now. I will probably restructuralize it a bit in the future :)

File: test/bindings/withState.test.tsx (hunk 2 above)
Comment (adam-stanek) [reply test/bindings/withState.test.tsx:25]: Ok, please write some `FIXME:` comment so that I can take a look at it later.

Comment (belaczek) [reply test/bindings/withState.test.tsx:25]: ✅

**Pull request merged**